api/datasets.go

```diff
@@ -197,14 +199,72 @@ func extensionToMimeType(ext string) string {
 	}
 }
 
-func (h *DatasetHandlers) listHandler(w http.ResponseWriter, r *http.Request) {
-	args := lib.ListParamsFromRequest(r)
-	args.OrderBy = "created"
+// snoop reads from an io.ReadCloser and restores it so it can be read again
+func snoop(body *io.ReadCloser) (io.ReadCloser, error) {
+	if body != nil && *body != nil {
+		result, err := ioutil.ReadAll(*body)
+		(*body).Close()
+
+		if err != nil {
+			return nil, err
+		}
+		if len(result) == 0 {
+			return nil, io.EOF
+		}
+
+		*body = ioutil.NopCloser(bytes.NewReader(result))
+		return ioutil.NopCloser(bytes.NewReader(result)), nil
+	}
+	return nil, io.EOF
+}
+
+var decoder = schema.NewDecoder()
+
+// UnmarshalParams deserialzes a lib req params stuct pointer from an HTTP
+// request
+func UnmarshalParams(r *http.Request, p interface{}) error {
+	// TODO(arqu): this should be set on the global decoder
+	// probably means the decoder should live in the API struct or somewhere similar
+	decoder.IgnoreUnknownKeys(true)
```

Comment:
Any suggestions on this, I want to avoid having it living in a global instance like object and passing everything around and just be a standalone method.

Reply:
Do we _need_ to ignore unknown keys? I'm guessing this is to deal with the fact that we have non-matching params allowed like `page` and `pageSize`, but I'd like to have unknown keys produce an error. If we need this for now, let's add a `TODO` to remove it in a follow up

In this case I do think it's fine to have as a global. It's just for type caching to get a perf bump at runtime. I'd suggest we move both `UmarshalParams` and the decoder declaration to `api/api.go`

Reply:
Currently there are lots of endpoints that either don't have those generic params, or we ignored them anyways like `fsi=true` on many endpoints.

Reply:
Also moving to api is a bit harder as we don't have the `api/api.go` `Server` struct available in `lib` nor should we. I'll move it to `lib/api.go` and add the `TODOs`.